test/trafficsplit/trafficsplit_test.go

```diff
@@ -0,0 +1,262 @@
+package trafficsplit
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/linkerd/linkerd2/testutil"
+)
+
+var TestHelper *testutil.TestHelper
+
+const ZeroRps = "0.0rps"
+
+func TestMain(m *testing.M) {
+	TestHelper = testutil.NewTestHelper()
+	os.Exit(m.Run())
+}
+
+type statTsRow struct {
+	name       string
+	apex       string
+	leaf       string
+	weight     string
+	success    string
+	rps        string
+	latencyP50 string
+	latencyP95 string
+	latencyP99 string
+}
+
+func parseStatTsRow(out string, expectedRowCount, expectedColumnCount int) (map[string]*statTsRow, error) {
+	rows, err := testutil.CheckRowCount(out, expectedRowCount)
+	if err != nil {
+		return nil, err
+	}
+
+	statRows := make(map[string]*statTsRow)
+
+	for _, row := range rows {
+		fields := strings.Fields(row)
+
+		if len(fields) != expectedColumnCount {
+			return nil, fmt.Errorf(
+				"Expected [%d] columns in stat output, got [%d]; full output:\n%s",
+				expectedColumnCount, len(fields), row)
+		}
+
+		row := &statTsRow{
+			name:       fields[0],
+			apex:       fields[1],
+			leaf:       fields[2],
+			weight:     fields[3],
+			success:    fields[4],
+			rps:        fields[5],
+			latencyP50: fields[6],
+			latencyP95: fields[7],
+			latencyP99: fields[8],
+		}
+
+		statRows[row.leaf] = row
+
+	}
+	return statRows, nil
+}
+
+func statTrafficSplit(from string, ns string) (map[string]*statTsRow, error) {
+	cmd := []string{"stat", "ts", "--from", from, "--namespace", ns}
+	stdOut, _, err := TestHelper.LinkerdRun(cmd...)
+	if err != nil {
+		return nil, err
+	}
+	return parseStatTsRow(stdOut, 2, 9)
+}
+
+func validateTrafficSplit(actual *statTsRow, expected *statTsRow) error {
+
+	if actual.name != expected.name && expected.name != "" {
+		return fmt.Errorf("expected name '%s' for leaf %s, got '%s'", expected.name, expected.leaf, actual.name)
+	}
+
+	if actual.apex != expected.apex && expected.apex != "" {
+		return fmt.Errorf("expected apex '%s' for leaf %s, got '%s'", expected.apex, expected.leaf, actual.apex)
+	}
+
+	if actual.weight != expected.weight && expected.weight != "" {
+		return fmt.Errorf("expected weight '%s' for leaf %s, got '%s'", expected.weight, expected.leaf, actual.weight)
+	}
+
+	if actual.success != expected.success && expected.success != "" {
+		return fmt.Errorf("expected success '%s' for leaf %s, got '%s'", expected.success, expected.leaf, actual.success)
+	}
+
+	if expected.rps != "-" {
```

Comment:
I'm having trouble following the nested ifs with non-equals.  I think this is saying...

if expected.rps is "" then pass
if expected.rps is "-" then actual must be "-"
if expected.rps is "0.0rps" then actual must be "0.0rps"
if expected.rps is a non-zero number, then actual must be a non-zero number

Is that right?  Can these be re-structured to be more straightforward?

Reply:
I really miss pattern matching in cases like this one